slapo/schedule.py

```diff
@@ -391,6 +398,11 @@ def validate_sync_op(mode, sync_op_or_fn, axis=None):
             elif mode == "fwd_post" and sync_op_or_fn == "reduce_scatter":
                 if output_type == "partition":
                     raise ValueError("Cannot reduce-scatter a partition output")
+            elif mode == "fwd_pre" and sync_op_or_fn == "all_gather":
+                if output_type == "partial":
```

Comment:
this is for assertion. if the layer is sharded in the input dimension, we cannot call allgather on the input

Reply:
just saying the parameter cannot be sharded is not accurate, since we allow the output dimension to be sharded. that's why i said "when the layer weight is partitioned in the **input dimension**"

Reply:
Still a bit confuse to me, but I don't want to block this PR because of an error message. We could refine it in the future.

Reply:
i have added more test. hope it is clearer